### PR TITLE
Add Clifford-only simulator (Stim)

### PR DESCRIPTION
**Pull request by bmhowe23**
Add https://github.com/quantumlib/Stim (v1.13.0) simulator for Clifford-only circuits.

The user requests it like other targets: `--target stim` and `cudaq.set_target('stim')`.

### CODE REVIEW

File: runtime/nvqir/stim/StimCircuitSimulator.cpp [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+  std::mt19937_64 randomEngine;
+
+  /// @brief Grow the state vector by one qubit.
+  void addQubitToState() override { addQubitsToState(1); }
+
+  /// @brief Override the default sized allocation of qubits
+  /// here to be a bit more efficient than the default implementation
+  void addQubitsToState(std::size_t qubitCount,
+                        const void *stateDataIn = nullptr) override {
+    return;
```
Comment (1tnguyen) [inline runtime/nvqir/stim/StimCircuitSimulator.cpp:37]: nit: should we throw an error here when `stateDataIn` is not null?

File: runtime/nvqir/stim/StimCircuitSimulator.cpp [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    std::transform(gateName.begin(), gateName.end(), gateName.begin(),
+                   ::toupper);
+    std::vector<std::uint32_t> stimTargets;
+
+    // These CUDA-Q rotation gates have the same name as Stim "reset" gates.
+    // Stim is a Clifford simulator, so it doesn't actually support rotational
+    // gates. Throw exceptions if they are encountered here.
+    if (gateName == "RX" || gateName == "RY" || gateName == "RZ")
+      throw std::runtime_error(fmt::format(
+          "Gate not supported by simulator: {}", task.operationName));
```
Comment (1tnguyen) [inline runtime/nvqir/stim/StimCircuitSimulator.cpp:86]: nit: should we add a TODO comment here about the possibility of handling `+/- pi/2` rotations for future references?

File: runtime/nvqir/stim/StimCircuitSimulator.cpp [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    stimCircuit.safe_append_u(
+        "R", std::vector<std::uint32_t>{static_cast<std::uint32_t>(index)});
+  }
+
+  /// @brief Sample the multi-qubit state.
+  cudaq::ExecutionResult sample(const std::vector<std::size_t> &qubits,
+                                const int shots) override {
+    std::vector<std::uint32_t> stimTargetQubits;
+    for (auto q : qubits)
+      stimTargetQubits.push_back(static_cast<std::uint32_t>(q));
```
Comment (1tnguyen) [inline runtime/nvqir/stim/StimCircuitSimulator.cpp:147]: nit
```suggestion
    std::vector<std::uint32_t> stimTargetQubits(qubits.begin(), qubits.end());
```

**Review by 1tnguyen — APPROVED** [review]
LGTM 👍

File: runtime/nvqir/stim/StimCircuitSimulator.cpp (hunk 1 above)
Comment (bmhowe23) [reply runtime/nvqir/stim/StimCircuitSimulator.cpp:40]: Updated in 20343fd07

File: runtime/nvqir/stim/StimCircuitSimulator.cpp (hunk 2 above)
Comment (bmhowe23) [reply runtime/nvqir/stim/StimCircuitSimulator.cpp]: Updated in 20343fd07

File: runtime/nvqir/stim/StimCircuitSimulator.cpp (hunk 3 above)
Comment (bmhowe23) [reply runtime/nvqir/stim/StimCircuitSimulator.cpp]: Updated in 20343fd07

**Comment by bmhowe23** [comment]
> These changes look great! Thanks, Ben. I don't see any new tests... It seems all of the C++ tests will be run except the ones which are skipped with `#ifndef CUDAQ_BACKEND_STIM`. What about Python - may be we should add a couple which showcase which gates are supported and behavior on unsupported gates.

Thanks for the suggestion, Pradnya. I added some Stim-specific backend tests in 20343fd07. Please take a look and let me know if that's what you had in mind, or if you'd like to see something else, too.

**Review by khalatepradnya — APPROVED** [review]
👍

**Pull request merged**